### ISSUE TITLE
nimble init shouldn't prompt by default

### ISSUE THREAD

**Issue by dom96** [nim-lang/nimble#288]
The ``init`` command should analyse the project and create an appropriate ``.nimble`` file with no prompts. The current behaviour should be behind a ``--prompt`` flag.

**Comment by antizealot1337** [comment]
I think I'd like to add this but don't know which way to go. Personally, I like the not prompting by default but figure it would be very unexpected for most users to get a new nimble version and `init` act very different.

I also have a question of the flag to change the behavior. I figure if not prompting `--prompt` would be the flag name. I'm not a big fan of `--no-prompt` for not prompting so how do you feel about `--guess`, `--auto`, or maybe even `--default`/`--defaults`? Another flag alternative could be `--prompt:on|off` which resembles how the compiler enables and disables checks. It may allow changing the default behavior of `init` in the future without much friction.

I'm also unclear how much analysis you would want nimble to do for an existing project?

**Comment by dom96** [comment]
> Personally, I like the not prompting by default but figure it would be very unexpected for most users to get a new nimble version and init act very different.

That's true, I think showing a "Hint: New version of Nimble no longer prompts for `init`. See <url> for more details" message would help the expectations.

I'm still leaning towards Nimble not prompting by default, so ``--prompt`` sounds good.  You could probably support the trifecta: ``--prompt``, ``--prompt:off``, ``--prompt:on`` :)

> I'm also unclear how much analysis you would want nimble to do for an existing project?

For a first version it's important to support the common case: you've just created a new directory called "myawesomeproject" and want to make it into a Nimble package. You can infer the same things that Nimble already infers: the project name (based on directory name) and author name (based on Git info IIRC).

You could also create a nice directory structure for the user:

* src/PkgName/
* src/PkgName.nim
* tests/sampletest.nim

Future versions would add support for existing projects being turned into a Nimble package (but this should be pretty rare)

**Comment by antizealot1337** [comment]
Awesome. Thanks for the advise. :-D

I'll go ahead and start working on this and get a PR as soon as possible.

**Comment by FedericoCeratto** [comment]
Nimble should still ask for the license!
Also #413 would require prompting for some options.
Removing the prompt by default sounds like surprising the user unnecessarily.
Plus, being shown reasonable guesses (e.g. project name, user name) and simply pressing enter takes very little time and people are starting new projects that frequently.

**Comment by antizealot1337** [comment]
@FedericoCeratto I agree this change would be surprising to Nimble users and got to thinking about how other languages and package managers do things. So, I did a little research (stuff on my machine and Google for things that weren't) to see how people trying Nim may feel about the behavior. My "research" is non-exhaustive as I picked mostly languages and package managers I've used at some point or another. Some of my information may be out of date and if so I apologize in advanced. I also would like to apologize for how long winded this post is. I didn't intend for it to be this long :/

These languages and their package managers have the current behavior as Nim and Nimble:
* D/DUB
* NodeJS/NPM
* Python<sup>1</sup>
* Scala/SBT<sup>2</sup>

These languages and their package managers do not prompt the user for any information:
* Rust/cargo
* Go<sup>3</sup>
* Crystal/shards
* Java/Maven

I think Go can be ignored and maybe Python too unless I'm mistaken (see notes). I'm not really familiar with Scala's SBT or Java's Maven so I can't really say how they work. In any case I see no clear precedent for a case of either prompting or not. Combined with the disruption of current users and possible deviation from the Nim in Action book may be a good argument for preserving current behavior. Though if Nimble offers good defaults (and I think it does) I think the chances of it negatively impacting users is low. Nimble could to provide a hint with at least one release saying "Next version will not prompt by default" and changing in a later release, but I don't know if that's strictly necessary.

Rust and Crystal are the only languages that don't prompt the user when `init` is used to create a new project and both appear to respect `license` as a key in their package files but don't set it by default. Crystal includes the key and sets it to MIT but comments it out. Rust completely omits it any license key. If following their lead it would make a good case for #419 and follow your suggestion or at least suggest that defaulting to a license isn't a bad thing. A `--license:mit` flag could be added to adjust, but may be more cumbersome than just editing the `.nimble` file after creation.

> ...add a commented-out version field or a special value to use the git tag.

As far as #413 is concerned I think using good defaults and a few command line flags to change the default behavior for those cases it doesn't work for wouldn't require prompting the user but I could be wrong.

I have no strong feeling for prompting or not *by default*. I would really like to see nimble offer the ability to create a package without prompting the user at all. It is slightly bothersome to create a package and have the package manager prompt information it already knows. I can hit enter to accept all the values until I get to description which must not be empty and sometimes isn't important but still required.

If it's determined the current behavior is to be maintained I'd be happy either fix the current PR or make a new one. :)

<sup>1</sup>: I googled "pip init" and found a program *pip-init* that prompts, but I don't know if `pip` has and `init` by default.
<sup>2</sup>: SBT has `sbt new` that appears to act like `init` but I don't know.
<sup>3</sup>: Go's current package manager is technically `go get` but I actually referenced the [dep](https://github.com/golang/dep) tool that is slated to be included as `go dep` at some unspecified time in the future. `dep init` creates a package file, but overall it doesn't feel that it and the other package managers' `init` have a lot in common.

**Comment by Araq** [comment]
Confusing IMHO. ``nimble -y`` exists to make nimble shut up and it should stay this way. Maybe name it ``--quiet``  or ``--prompt:off`` I don't care. But make it consistent, Nimble is curently chatty anyway.

**Comment by antizealot1337** [comment]
I forgot about `-y`. I haven't used it much. I looked into it and don't see any issues overloading `-y` in `init` and just use all the default values. I'll make another PR since the other is larger and this seems fairly trivial anyway.

**Comment by dom96** [comment]
Thanks for the thorough research. So perhaps we should change what is prompted by default, i.e. only things that we definitely don't know like the license. The name of the project, author etc. is readily available information so hitting enter for it each time is annoying.

What I would like to see for the license is a drop-down style selection. A set of choices you can scroll through using the arrow keys in the terminal. This might require some magic but would be really nice.

And of course we can have a switch that just selects the defaults, and for that reusing ``-y`` sounds good.

**Comment by antizealot1337** [comment]
No problem. I figured it couldn't hurt to see how other people dealt with package creation. Would there be any scenario you would prompt for the *package*? I imagine if *author* can't be determined nimble should prompt the user then. I'd still like to create the directory structure too.

Drop down like behavior would be really nice. I'd like to try to implement it. I imagine that would require `terminal` and `isatty` may return false, so it would require a fallback behavior too. I imagine fallback would just display a prompt like now with all the valid options and validate the input. I can start with the fallback and push while play with the other. Would the MIT, BSD2, GPL3, LGPL3, and Apache2 licenses be an adequate selection?

I realize I probably jumped too fast with the new PR without waiting for direction. I'll definitely wait more in the future for more consensus. :)
